pkg/ddl/job_table.go

```diff
@@ -507,93 +510,123 @@ func (d *ddl) delivery2LocalWorker(pool *workerPool, task *limitJobTask) {
 // delivery2Worker owns the worker, need to put it back to the pool in this function.
 func (s *jobScheduler) delivery2Worker(wk *worker, pool *workerPool, job *model.Job) {
 	injectFailPointForGetJob(job)
-	s.runningJobs.add(job)
-	s.wg.Run(func() {
-		metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Inc()
+	jobID, involvedSchemaInfos := job.ID, job.GetInvolvingSchemaInfo()
+	s.runningJobs.add(jobID, involvedSchemaInfos)
+	metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Inc()
+	s.wg.RunWithLog(func() {
 		defer func() {
+			pool.put(wk)
 			failpoint.InjectCall("afterDelivery2Worker", job)
-			s.runningJobs.remove(job)
+			s.runningJobs.remove(jobID, involvedSchemaInfos)
 			asyncNotify(s.ddlJobNotifyCh)
 			metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Dec()
 			if wk.ctx.Err() != nil && ingest.LitBackCtxMgr != nil {
```

Comment:
At this point of time, `wk` is put back into the pool and may be get by someone else. But it is still used by here, will there be any problems?

Reply:
wk.ctx is same as scheduler.schCtx now, it should be ok，but let's move `push back` after last usage

Reply:
https://github.com/pingcap/tidb/pull/53747/commits/1e89963a3403428fb7df4ac91fa6eb889e91ac77